src/viur/core/modules/user.py

```diff
@@ -616,133 +619,211 @@ def login(self, skey="", token="", *args, **kwargs):
 
 
 class TimeBasedOTP:
-    windowSize = 5
+    WINDOW_SIZE = 5
+    MAX_RETRY = 3
     otpTemplate = "user_login_timebasedotp"
 
+    @dataclasses.dataclass
+    class OtpConfig:
+        """
+        This dataclass is used to provide an interface for a OTP token
+        algorithm description that is passed within the TimeBasedOTP
+        class for configuration.
+        """
+        secret: str
+        timedrift: float = 0.0
+        algorithm: typing.Literal["sha1", "sha256"] = "sha1"
+        interval: int = 60
+
+    class OtpSkel(skeleton.RelSkel):
+        """
+        This is the Skeleton used to ask for the OTP token.
+        """
+        otptoken = NumericBone(
+            descr="Token",
+            required=True,
+            max=999999,
+            min=0,
+        )
+
     def __init__(self, userModule, modulePath):
         super().__init__()
         self.userModule = userModule
         self.modulePath = modulePath
 
     @classmethod
-    def get2FactorMethodName(*args, **kwargs):
+    def get2FactorMethodName(*args, **kwargs):  # fixme: What is the purpose of this function? Why not just a member?
         return "X-VIUR-2FACTOR-TimeBasedOTP"
 
-    def canHandle(self, userKey) -> bool:
-        user = db.Get(userKey)
-        return all(
-            [(x in user and (x == "otptimedrift" or bool(user[x]))) for x in ["otpid", "otpkey", "otptimedrift"]])
-
-    def startProcessing(self, userKey):
-        user = db.Get(userKey)
-        if all([(x in user and user[x]) for x in ["otpid", "otpkey"]]):
-            logging.info("OTP wanted for user")
-            session = current.session.get()
-            session["_otp_user"] = {
-                "uid": str(userKey),
-                "otpid": user["otpid"],
-                "otpkey": user["otpkey"],
-                "otptimedrift": user["otptimedrift"],
-                "timestamp": time.time(),
-                "failures": 0
-            }
-            session.markChanged()
-            return self.userModule.render.loginSucceeded(msg="X-VIUR-2FACTOR-TimeBasedOTP")
+    def get_config(self, user_key) -> OtpConfig | None:
+        """
+        Returns an instance of self.OtpConfig with a provided token configuration,
+        or None when there is no appropriate configuration of this second factor handler available.
+        """
+        user = db.Get(user_key)
+        if user and all(bool(user.get(k)) for k in ("otp_serial", "otp_secret")):
```

Comment:
I think there's no reason to enforce the `otp_serial`.
```suggestion
        if user and user.get("otp_secret"):
```

Reply:
Indeed!